pinot-core/src/main/java/com/linkedin/pinot/core/data/readers/ThriftRecordReader.java

```diff
@@ -0,0 +1,148 @@
+
+    private final Schema _schema;
+
+    private final ThriftRecordReaderConfig _recordReaderConfig;
+
+    private final Class<TBase<?, ?>> _thriftClass;
+
+    /**
+     * For reading the binary thrift objects.
+     */
+    private TBinaryProtocol _binaryIn;
+
+    private BufferedInputStream _bufferIn;
+
+    private final File _dataFile;
+
+    Map<String, Integer> _fieldNameToIndexMap;
+
+    public ThriftRecordReader(File dataFile, Schema schema, ThriftRecordReaderConfig recordReaderConfig) throws IOException, ClassNotFoundException, IllegalAccessException, InstantiationException {
+        this._schema = schema;
+        this._dataFile = dataFile;
+        this._recordReaderConfig = recordReaderConfig;
+        this._thriftClass = thriftClassCreator();
+        this._bufferIn = RecordReaderUtils.getFileBufferStream(dataFile);
+        this._binaryIn = new TBinaryProtocol(new TIOStreamTransport(_bufferIn));
+        this._fieldNameToIndexMap = new HashMap();
+        TBase t = this._thriftClass.newInstance();
+        int index = 1;
+        TFieldIdEnum fieldIdEnum = null;
+        do {
+            fieldIdEnum = t.fieldForId(index);
+            if (fieldIdEnum != null) {
+                _fieldNameToIndexMap.put(fieldIdEnum.getFieldName(), index);
+            }
+            index = index + 1;
+        } while (fieldIdEnum != null);
+    }
+
+    @Override
+    public boolean hasNext() {
+        _bufferIn.mark(1);
+        int val = 0;
+        try {
+            val = _bufferIn.read();
+            _bufferIn.reset();
+        } catch (IOException e) {
+            val = -1;
+            throw new RuntimeException("Error in iterating Reader", e);
+        }
+        return val != -1;
+    }
+
+    @Override
+    public GenericRow next() throws IOException {
+        return next(new GenericRow());
+    }
+
+    @Override
+    public GenericRow next(GenericRow reuse) throws IOException {
+        TBase t = null;
+        try {
+            t = this._thriftClass.newInstance();
+            t.read(_binaryIn);
+        } catch (Exception e) {
+            throw new RuntimeException("Caught exception while serialize thrift instance", e);
+        }
+        for (FieldSpec fieldSpec : _schema.getAllFieldSpecs()) {
+            String fieldName = fieldSpec.getName();
+            if (_fieldNameToIndexMap.containsKey(fieldName)) {
+                int tFieldId = _fieldNameToIndexMap.get(fieldName);
+                TFieldIdEnum tFieldIdEnum = t.fieldForId(tFieldId);
+                Object thriftValue = t.getFieldValue(tFieldIdEnum);
+                Object value;
+                if (fieldSpec.isSingleValueField()) {
+                    String token = thriftValue != null ? thriftValue.toString() : null;
+                    value = RecordReaderUtils.convertToDataType(token, fieldSpec);
+                } else {
+                    value = RecordReaderUtils.convertToDataTypeArray((ArrayList) thriftValue, fieldSpec);
```

Comment:
will thriftValue always be an ArrayList?

Reply:
Thrift support `ArrayList`, `HashSet` and `HashMap`. I will add support for `HashSet` and exclude `HashMap` for now.